### DISCUSSION TITLE
Is it possible to build static library of libOnnxruntime if somebody is aware? We are stuck. As per our business use case we can not use .so library.  Feeling really stuck on the same.

### DISCUSSION THREAD

**Discussion by dsaneja** [microsoft/onnxruntime#16799] [Other Q&A]
Is it possible to build static library of libOnnxruntime if somebody is aware? We are stuck. As per our business use case we can not use .so library.

Feeling really stuck on the same.

**Comment by xadupre** [comment]
This option is not available for all the configuration but it is possible to statically build onnxruntime: https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/build.py#L286C28-L286C44. The flag `--build_shared_lib` should not be part of the command line you are using the build onnxruntime.

**Reply by dsaneja** [reply]
You must be highlighting the python one. I want to build the c++ one of onnxruntime.

**Comment by xadupre** [comment]
The script is written in python but it is used to build any version of onnxruntime. The python build is enabled with option --build_wheel. If not present, python bindings is not built. It builds the full command line for cmake and then calls cmake.

**Reply by dsaneja** [reply]
The problem is that it creates lot of .a(static ) files/individual  instead of one single static file. There should be one .a file as like .so

**Comment by xadupre** [comment]
It would require to change the cmake file to produce one file or maybe you can insert the command you need in the makefile to produce one. The list you need is defined here: https://github.com/microsoft/onnxruntime/blob/main/cmake/onnxruntime.cmake#L195.

**Comment by xeeetu** [comment]
@xadupre I managed to create a static onnxruntime library. Can you please tell me if it is possible to create a static version of onnxruntime with DirectML EP? If it's so, any hints would help me a lot.

**Reply by dsaneja** [reply]
@xeetu : Can you please provide me the static build library somewhere. I shall be grateful to you.

**Reply by dsaneja** [reply]
any box link or google drive link whatever works best for you. Because I tried building onnxruntime library statically but no luck with me & my team so far.

**Comment by xadupre** [comment]
I assume you added `--use_dml` when running the command line or `-Donnxruntime_USE_DML` on the cmake command line. The options related to winml are written in `cmake/winml.cmake`. I assume this was written to link with a dll but it should be possible to modify it to link with a static library.

**Comment by PenutChen** [comment]
I'm curious, in what scenario would you be unable to use a .so file?
